File: solutions/haiku-1/haiku/ruby/roman.rb
def to_roman(n)
  # Mapping of values to Roman numerals in descending order
  # Includes subtractive forms: IV, IX, XL, XC, CD, CM
  values = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
  numerals = ['M', 'CM', 'D', 'CD', 'C', 'XC', 'L', 'XL', 'X', 'IX', 'V', 'IV', 'I']

  result = ""

  values.each_with_index do |value, index|
    while n >= value
      result += numerals[index]
      n -= value
    end
  end

  result
end

# Read input and convert each number
while line = STDIN.gets
  n = line.strip.to_i
  puts to_roman(n)
end

File: solutions/haiku-1/haiku/java/roman.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Define the mapping of values to roman numerals in descending order
        // Includes subtractive forms: IV, IX, XL, XC, CD, CM
        int[] values = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] numerals = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

        while (scanner.hasNextLine()) {
            int n = Integer.parseInt(scanner.nextLine());
            StringBuilder roman = new StringBuilder();

            // Greedily subtract the largest possible value and append the symbol
            for (int i = 0; i < values.length; i++) {
                while (n >= values[i]) {
                    roman.append(numerals[i]);
                    n -= values[i];
                }
            }

            System.out.println(roman.toString());
        }
    }
}
